### ISSUE TITLE
CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered.

### ISSUE THREAD

**Issue by nectario** [tensorflow/tensorflow#50735]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**This is an urgent issue! It has caused a production problem!**

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux (Amazon Sagemaker)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): **latest**: 2.5
- Python version: 3.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.020/8.202
- GPU model and memory: Tesla T4/13.8 GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I updated my model and when I train, I get the following error:

`
2021-07-11T23:08:22.514-04:00	2021-07-12 03:08:22.053193: W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:689] Error in PredictCost() for the op: op: "Softmax" attr { key: "T" value { type: DT_FLOAT } } inputs { dtype: DT_FLOAT shape { unknown_rank: true } } device { type: "GPU" vendor: "NVIDIA" model: "Tesla T4" frequency: 1590 num_cores: 40 environment { key: "architecture" value: "7.5" } environment { key: "cuda" value: "11020" } environment { key: "cudnn" value: "8100" } num_registers: 65536 l1_cache_size: 24576 l2_cache_size: 4194304 shared_memory_size_per_multiprocessor: 65536 memory_size: 14474280960 bandwidth: 320064000 } outputs { dtype: DT_FLOAT shape { unknown_rank: true } }

2021-07-11T23:08:22.514-04:00	2021-07-12 03:08:22.053330: W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:689] Error in PredictCost() for the op: op: "Softmax" attr { key: "T" value { type: DT_FLOAT } } inputs { dtype: DT_FLOAT shape { unknown_rank: true } } device { type: "GPU" vendor: "NVIDIA" model: "Tesla T4" frequency: 1590 num_cores: 40 environment { key: "architecture" value: "7.5" } environment { key: "cuda" value: "11020" } environment { key: "cudnn" value: "8100" } num_registers: 65536 l1_cache_size: 24576 l2_cache_size: 4194304 shared_memory_size_per_multiprocessor: 65536 memory_size: 14474280960 bandwidth: 320064000 } outputs { dtype: DT_FLOAT shape { unknown_rank: true } }

2021-07-11T23:08:32.516-04:00	2021-07-12 03:08:32.060022: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudnn.so.8

2021-07-11T23:08:34.517-04:00	2021-07-12 03:08:34.183639: I tensorflow/stream_executor/cuda/cuda_dnn.cc:359] Loaded cuDNN version 8202

2021-07-11T23:08:36.517-04:00	2021-07-12 03:08:36.256607: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.11

2021-07-11T23:08:38.518-04:00	2021-07-12 03:08:38.161603: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.11

2021-07-11T23:08:50.608-04:00

**2021-07-12 03:08:50.058288: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
2021-07-12 03:08:50.058288: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountere**d`

**Describe the expected behavior**

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**LOGS:**
[logs.txt](https://github.com/tensorflow/tensorflow/files/6801798/logs.txt)

**Comment by tilakrayal** [comment]
@nectario ,

 Can you please try installing TensorFlow v2.5 with CUDA 11.2 and cuDNN 8.1 and check if you are facing the same error.For more information please take a look at the [tested build configurations](https://www.tensorflow.org/install/source#gpu). Thanks!

**Comment by nectario** [comment]
> @nectario ,
> 
> Can you please try installing TensorFlow v2.5 with CUDA 11.2 and cuDNN 8.1 and check if you are facing the same error.For more information please take a look at the [tested build configurations](https://www.tensorflow.org/install/source#gpu). Thanks!

Thank you. Running on my my local machine, where I have these versions, it works fine. But the real issue is that I am using Amazon Sagemaker and that build is fixed. The Docker image I am pulling is tensorflow/tensorflow:latest-gpu-jupyter 
I also tried pulling the Nvidia image of tensorflow from nvcr.io/nvidia/tensorflow:21.06-tf2-py3 
but I still have the same problem. This happens with the specific deep learning architecture I am using. My older architecture does not have this problem.

**Comment by nectario** [comment]
This issue has completely derailed our production release. I hope it's given some urgency.

**Comment by nectario** [comment]
Is there a tensorflow docker image where I can specify the CUDA/CUDNN version specifically?

From my Dockerfile I am calling:

`FROM tensorflow/tensorflow:latest-gpu-jupyter`

**Comment by nectario** [comment]
From the logs, I see it's loading CUDNN version 8.2:

`2021-07-13 21:21:35.642481: I tensorflow/stream_executor/cuda/cuda_dnn.cc:359] Loaded cuDNN version 8202`

**Comment by nectario** [comment]
This error also happens with the tf-nightly image.

**Comment by nectario** [comment]
I got it to work when I use the tensorflow Docker image for TF 2.4.2: 

tensorflow/tensorflow:2.4.2-gpu-jupyter

So it is clearly a problem with later versions of CUDA/CUDNN!

**Comment by nectario** [comment]
This is the Google Colab with the code that triggers this error. If you run this here, it will not happen in the colab notebook as it happens mainly in the Docker image tensorflow/tensorflow:latest-gpu-jupyter

Code:

https://colab.research.google.com/drive/1E_IZ3zmPrQHbuiYl47gvL591hHfS-VkN#scrollTo=ts7l0877A0_d

**Comment by nectario** [comment]
Is there any update on this? I really would like to use version 2.6, but cannot.

**Comment by sanjoy** [comment]
Hi @nectario,

Can you try running with `CUDA_LAUNCH_BLOCKING=1` in the environment & share the logs?  That should isolate the error to the actual GPU operation that caused it.

Also: do you have a reproducer that does not depend on separate data files, like `./y_val_two_week_return.pkl`?

**Comment by nectario** [comment]
Hi @sanjoy 

Sorry for my late response. I just came back from vacation. I did as you said, but surprisingly, the issue is not reproduceable now. I tried running it in the different environments in AWS and also my local machine, where it was happening, and it does not happen now. The version of tensorflow I tested with  was 2.6. Fyi, when this first happened, it was under 2.5. Anyway, I do remember testing also with 2.6 back then, and it was happening when 2.6 first came out.

If you want me to run it with 2.5 with the above setting, let me know.

**Comment by nectario** [comment]
@sanjoy I take it back. This error did manage to happen again. Surprisingly it happened when I removed the above flag CUDA_LAUNCH_BLOCKING=1. From my 250 model training 13 succeeded but the rest all failed with that error. Unfortunately it was without the flag. Will run it  with it and will attach the logs.

**Comment by nectario** [comment]
@sanjoy Now that I think about it, the day before yesterday I trained my 250 models with the flag on and the error did not happen. Although training seemed a little bit slower than usual. And today I trained with the flag removed, and 200+ model trainings failed.

**Comment by utkarshgupta137** [comment]
@sanjoy this flag helped me as well, although I only found it after 1 week of head-scratching.

Config:
Ubuntu 20.04.3 LTS
Docker 20.10.8
Image: tensorflow/tensorflow:2.6.0-gpu-jupyter
GPU: Telsa P40 (Azure VM)
Drivers: 470.57.02
CUDA: 11.4

**Comment by nectario** [comment]
This problem continues to happen and it only stops happening if I include the env flag: CUDA_LAUNCH_BLOCKING=1

**Comment by sanjoy** [comment]
> This problem continues to happen and it only stops happening if I include the env flag: CUDA_LAUNCH_BLOCKING=1

Unfortunately this means this is a heisenbug. :(

Running with `CUDA_LAUNCH_BLOCKING=1` "for real" is not a good idea since it affects performance, as you indicated in https://github.com/tensorflow/tensorflow/issues/50735#issuecomment-917314918.

Will you be able to extract a minimal reproducer that shows the issue?  Ideally something that only has a few ops?

**Comment by nectario** [comment]
I just emailed you a file with the code that reproduces this issue on a Linux box. 

> > This problem continues to happen and it only stops happening if I include the env flag: CUDA_LAUNCH_BLOCKING=1
> 
> Unfortunately this means this is a heisenbug. :(
> 
> Running with `CUDA_LAUNCH_BLOCKING=1` "for real" is not a good idea since it affects performance, as you indicated in [#50735 (comment)](https://github.com/tensorflow/tensorflow/issues/50735#issuecomment-917314918).
> 
> Will you be able to extract a minimal reproducer that shows the issue? Ideally something that only has a few ops?

**Comment by rsnitsch** [comment]
I get the same error on AWS instances running in my EKS cluster.

My environment:
- instance type: p2.xlarge (having K80 GPUs)
- architecture: linux (amd64)
- AMI: 0dec13e4e8336258b ([Amazon EKS optimized accelerated Amazon Linux AMI](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html#gpu-ami), it is based on Amazon Linux 2)
- Kernel version: 5.4.188-104.359.amzn2.x86_64
- Tensorflow 2.8.0
- nvidia-smi output: NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4
- cuDNN version 8400
- container runtime: docker 20.10.13
- kubelet version: v1.22.6-eks

**Comment by rsnitsch** [comment]
Addendum: I just tested the same container image on a p3.2xlarge instance and the error did **not** occur. To confirm this I switched back to p2 and the error reappeared...

**Comment by cosminc98** [comment]
> Addendum: I just tested the same container image on a p3.2xlarge instance and the error did **not** occur. To confirm this I switched back to p2 and the error reappeared...

Can you provide the information on the environment in which it did work (the p3.2xlarge)? Information like what image is the container based on and/or what CUDA/tensorflow versions you have.

**Comment by rsnitsch** [comment]
> > Addendum: I just tested the same container image on a p3.2xlarge instance and the error did **not** occur. To confirm this I switched back to p2 and the error reappeared...
>
> @rsnitsch
> Can you provide the information on the environment in which it did work (p3.2xlarge)? Information like what image is the container based on and/or what CUDA/tensorflow versions you have.

I described the environment in detail in this comment above: https://github.com/tensorflow/tensorflow/issues/50735#issuecomment-1118669114

**Comment by foxik** [comment]
This might be #60192, which shows that since TF 2.5, masked cuDNN GRU/LSTM operations randomly crash and also give incorrect results (when they do not crash). The culprit seems to be (at least to me) the cuDNN 8.1 codepath; when cuDNN 8.0 or older is used, or when no masking is used (so not ragged tensors nor masked elements), everything works fine.

**Comment by SuryanarayanaY** [comment]
Hi @nectario ,

Can you confirm whether this issue is duplicate of 60192 as mentioned in above [comment](https://github.com/tensorflow/tensorflow/issues/50735#issuecomment-1502408785) ? If so we can close it here and track the issue there as it is active there.

If not can you please check and confirm with latest versions ? If not we can escalate this issue to concerned. Thanks!

**Comment by naseemap47** [comment]
> @sanjoy Now that I think about it, the day before yesterday I trained my 250 models with the flag on and the error did not happen. Although training seemed a little bit slower than usual. And today I trained with the flag removed, and 200+ model trainings failed.

Hi @nectario ,
I also faced the same issue. But I fix this by giving proper CUDA version for Tensorflow.
I am using **tensorflow==2.7.0 and CUDA 11.2**
I think this will help.
Thank You.

**Comment by foxik** [comment]
Hi all, the original problem causing cuDNN implementation of masked LSTM/GRU to fail with the `CUDA_ERROR_ILLEGAL_ADDRESS` on TF 2.5+ has been solved in cuDNN 8.9.2. So if you upgrade cuDNN to this version (you can use official TF build, even if it is build for cuDNN 8.2, the CUDA/cuDNN are backward compatible within the same major version), it should be fine -- see #60192.

**Comment by sainisatish** [comment]
I did a very small mistake , I have opened the python kernel instead of pytorch/tensorflow kernel in JupyterLab and got the same error.
